### PR TITLE
ref(*): refactor to act as jnlp-slave

### PR DESCRIPTION
**Pull request by vdice**
For use with the Kubernetes Jenkins Plugin.

TODO
 - [x] add/modify docs
 - [x] ~~either~~ update chart ~~dir or remove (if chart removed, external ci updates will be needed)~~

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,10 @@
-FROM ubuntu:16.04
+FROM openjdk:8-jdk
+
+ARG VERSION=3.7
```
Comment (bacongobbler) [inline Dockerfile]: can we change this to JENKINS_VERSION for clarity?

File: Dockerfile [hunk 2]
```diff
@@ -82,16 +88,14 @@ ENV SHELLCHECK_VERSION=0.4.3
 RUN curl -L https://s3-us-west-2.amazonaws.com/get-deis/shellcheck-$SHELLCHECK_VERSION-linux-amd64 -o /usr/local/bin/shellcheck \
     && chmod +x /usr/local/bin/shellcheck
 
-# copy everything to rootfs
-COPY rootfs /
-
 # change ownership of everything in $JENKINS_HOME to jenkins
 RUN chown -R jenkins:jenkins $JENKINS_HOME
 
+COPY rootfs/bin/jenkins-slave /usr/local/bin/jenkins-slave
```
Comment (bacongobbler) [inline Dockerfile]: How about we keep this as `COPY rootfs /` and dump the jenkins-slave script in `rootfs/usr/local/bin`? That way we can dump things in rootfs and it'll get dumped into the container's root filesystem. Makes it easier to add new things in the future.

**Review by bacongobbler — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (vdice) [reply Dockerfile]: will do (technically it's the Jenkins [Remoting](https://github.com/jenkinsci/remoting) Module version; will update accordingly)

**Comment by bacongobbler** [comment]
Once docs are added I'll LGTM this :)

**Comment by vdice** [comment]
@bacongobbler docs added and chart updated!  PTAL

**Comment by krancour** [comment]
I'm wondering if it doesn't make more sense to use `jenkinsci/jnlp-slave` as the base image. Wouldn't that spare us from having to include out own startup script?

**Comment by vdice** [comment]
> I'm wondering if it doesn't make more sense to use jenkinsci/jnlp-slave as the base image

It would be preferable if we could.  However, the current Dockerfile as written does quite a few root-user tasks (apt-get install *, mkdir, chown, etc.) and the `jenkinsci/jnlp-slave` image has `USER jenkins` set.  If there is a way to reconfig our Dockerfile to be able to do all it needs to do but as the `jenkins` user, then I'd definitely be in favor.  (Also note some custom logic in the `jenkins-slave` script)

**Comment by bacongobbler** [comment]
Is it possible to do this?

```
FROM jenkins/jnlp-slave
USER root
RUN apt-get install ...
USER jenkins
```

?

**Comment by bacongobbler** [comment]
> (Also note some custom logic in the jenkins-slave script)

We can override the jenkins-slave script provided in jenkins/jnlp-slave using ADD/COPY as usual. I had to do the same for deis/postgres and its docker-entrypoint script.

**Comment by bacongobbler** [comment]
Either way we can refactor in the future. LGTM!

**Review by bacongobbler — APPROVED** [review]

**Comment by vdice** [comment]
Great suggestions @krancour and @bacongobbler .  Pushed the latest changes and tested it via the following pipeline: https://ci.deis.io/job/builder-clone/1  (*edit*: sorry, deleted that clone job; here's the e2e run though: https://ci.deis.io/job/workflow-chart-e2e/33/)

I'd like to get our k8s plugin pulling this image and running more jobs just to be sure before merging, but this does simplify things.

**Comment by vdice** [comment]
Sigh; n/m... I was fooled by the pipeline b/c although the builder clone job using my agent succeeded, the downstream job was still using the previous jenkin-node:canary image.  Now when I flipped ci to use this test img, jobs fail w/ perms issues.  It seems we still need the container user to be root -- but would be great to refactor later.  I'll update this PR and keep testing...

**Comment by vdice** [comment]
Ok, we're back to where this PR originally started.  In order to refactor the image to pull from `jenkinsci/jnlp-slave` and not use our custom `jenkins-slave` script, we'll need to make updates in the `deis/e2e-runner` project.  I'll file a ticket for this work.

**Review by krancour — APPROVED** [review]

**Pull request merged**